### PR TITLE
Feature: Addition of Venue for Events

### LINKED ISSUE #1640: API - Add support for event venues
**Is your feature request related to a problem? Please describe.**

We need the ability to add API functionality to assign a venue to an event. Currently you can assign geographic coordinates, but you cannot also provide one of a list of pre-defined venues (eg. "Conference Room", "Sports Field" or "Theater") This would be a useful capability to add.

**Describe the solution you'd like**

The aim would be for Admins to do CRU activities (no deletions, but disabling/enabling) for venues on an Org screen, and for these venues to appear as options whenever events are scheduled.

The API must be able to:

1. determine whether the venue is already allocated during all or some part of the proposed event timeslot when the event is being created and not make it appear as an option.
    1. if there are no venues configured, or no available venues, then the returned result be meaningful and the app must not crash
2. reject the creation of an event if the requested location is unavailable
3. return an availability status for a selected venue for a given timeslot that could be used whenever the event timing is adjusted.

Other

1. We will also need the ability to set the maximum capacity of the venue in terms of user checkins. This will help with events that require pre-registration
1. Venues can often be reconfigured to have less or more seating. This means that there will need to be the ability to override the checkin limit on a per event basis.

The schema must include support for:

1. Venue ID
2. Venue name
3. Venue description 

This issue should be split into multiple PRs for ease of review:

1. Venue CRUD
2. Adding a venue to a single Event
3. Verifying the venue isn't already used for a single event
4. Managing event checkins for venue capacity
5. Adding the venues to recurring events

**Describe alternatives you've considered**
- N/A

**Approach to be followed (optional)**
- See above

**Additional context**

Related issues:
- https://github.com/PalisadoesFoundation/talawa-api/issues/1640
- https://github.com/PalisadoesFoundation/talawa-admin/issues/1390

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship 
- https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by wingman47**
**What kind of change does this PR introduce?**

Introduces venues for events. 
For now, it adds the capability to CRUD venues in organization and their respective tests. 

Subsequent PRs (as requested) will add other features like booking venues for events, avoid clashes of venues, managing event checkins for venue capacity.

**Issue Number:**

Fixes #1640

**Did you add tests for your changes?**

Yes.

**Snapshots/Videos:**

PART - 1

https://github.com/PalisadoesFoundation/talawa-api/assets/49762303/3fa2a010-292c-4894-a440-d30d13c74a7b

PART - 2

https://github.com/PalisadoesFoundation/talawa-api/assets/49762303/756fbbda-dd48-4b87-acfe-fe73b8fb3bdc

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by Cioppolo14** [comment]
@wingman47 Please fix the failing tests, the codecov and introspection tests should pass.

**Comment by wingman47** [comment]
> @wingman47 Please fix the failing tests, the codecov and introspection tests should pass. 

Working on it

**Comment by wingman47** [comment]
> @wingman47 Please fix the failing tests, the codecov and introspection tests should pass.

The failing codecov tests are not related to my changes. 
Only image upload is not tested in my files because I am not sure how to implement it. Any guidance would be helpful.
image


The workflow errors would be there due to the type of PR I think.

**Comment by palisadoes** [comment]
@wingman47 

1. We had to revert a PR recently. Please merge your code with the latest upstream.
2. Try to get the code coverage for you patch to pass

File: src/typeDefs/inputs.ts [hunk 1]
```diff
@@ -64,6 +64,15 @@ export const inputs = gql`
     name_of_user_starts_with: String
   }
 
+  input EditVenueInput {
+    _id: ID!
+    capacity: Int!
+    name: String!
```
Comment (EshaanAgg) [inline src/typeDefs/inputs.ts:70]: These should be allowed to be `undefined`, as it is not compulsory for the user to provide them if they aren't changed.

File: src/typeDefs/mutations.ts [hunk 2]
```diff
@@ -111,10 +111,16 @@ export const mutations = gql`
 
     createSampleOrganization: Boolean! @auth
 
+    createVenue(data: VenueInput!, file: String): Venue! @auth
```
Comment (EshaanAgg) [inline src/typeDefs/mutations.ts:114]: What is `file` here? Why isn't it included in the `data` itself?

File: src/typeDefs/mutations.ts [hunk 3]
```diff
@@ -111,10 +111,16 @@ export const mutations = gql`
...
+
     deleteAdvertisementById(id: ID!): DeletePayload!
 
     deleteDonationById(id: ID!): DeletePayload!
 
+    deleteVenue(venueId: ID!, organizationId: ID!): Venue @auth
```
Comment (EshaanAgg) [inline src/typeDefs/mutations.ts:120]: No need for `organizationId` here.

File: src/typeDefs/mutations.ts [hunk 4]
```diff
@@ -111,10 +111,16 @@ export const mutations = gql`
...
+
+    editVenue(data: EditVenueInput!, file: String): Venue! @auth
```
Comment (EshaanAgg) [inline src/typeDefs/mutations.ts:122]: Similar comments about `file`.

File: src/typeDefs/types.ts [hunk 5]
```diff
@@ -168,6 +168,7 @@ export const types = gql`
     status: Status!
     feedback: [Feedback!]!
     averageFeedbackScore: Float
+    venue: ID
```
Comment (EshaanAgg) [inline src/typeDefs/types.ts:171]: It should be returning a `Venue` with a field-level resolver.

File: src/typeDefs/queries.ts [hunk 6]
```diff
@@ -119,5 +119,7 @@ export const queries = gql`
       skip: Int
       orderBy: UserOrderByInput
     ): [User]! @auth
+
+    venuesInOrganization(id: ID!): [Venue] @auth
```
Comment (EshaanAgg) [inline src/typeDefs/queries.ts:123]: No need for this. Add `venues` as a field on the organization model, and then use the field level resolver for the same.

File: src/typeDefs/inputs.ts [hunk 7]
```diff
@@ -64,6 +64,15 @@ export const inputs = gql`
...
+    description: String
+    organizationId: ID!
```
Comment (EshaanAgg) [inline src/typeDefs/inputs.ts:72]: No need to provide this here.

File: src/typeDefs/types.ts [hunk 8]
```diff
@@ -298,6 +299,14 @@ export const types = gql`
...
   }
 
+  type Venue {
+    _id: ID!
+    name: String!
+    capacity: Int!
+    description: String
+    imageUrl: URL
+  }
+
```
Comment (EshaanAgg) [inline src/typeDefs/types.ts:309]: Add a field known as `organization` on the here, and implement a field level resolver for the same.

File: src/models/Organization.ts [hunk 9]
```diff
@@ -171,6 +174,12 @@ const organizationSchema = new Schema(
         ref: "CustomField",
       },
     ],
+    venues: [
+      {
+        type: Schema.Types.ObjectId,
+        ref: "Venue",
+      },
+    ],
```
Comment (EshaanAgg) [inline src/models/Organization.ts:182]: No need to store this information on the organization model.

File: src/models/Venue.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+  },
+  capacity: {
+    type: Number,
+    required: true,
+  },
+  imageUrl: {
+    type: String,
+    required: false,
+  },
+});
```
Comment (EshaanAgg) [inline src/models/Venue.ts:36]: Also store to which organization does this venue belong to here with a `organizationId` parameter.

**Review by EshaanAgg — CHANGES_REQUESTED** [review]
Needs some changes in the schema and the typedefs. When the same is completed, the changes must be migrated to the resolvers and the queries. 

Good job, though.

File: src/typeDefs/mutations.ts (hunk 3 above)
Comment (wingman47) [reply src/typeDefs/mutations.ts:120]: `organizationId` is being used here to check here if the user who want to delete the venue is admin of the organization or not. Should I take any other approach?

**Comment by wingman47** [comment]
> Needs some changes in the schema and the typedefs. When the same is completed, the changes must be migrated to the resolvers and the queries.
> 
> Good job, though.
Thanks.
I have made the required changes. I will modify the tests once the changes are confirmed. I hope its fine!

**Comment by EshaanAgg** [comment]
You also need to create a resolver for Organization on Venue type. The rest looks good.

**Comment by wingman47** [comment]
> You also need to create a resolver for Organization on Venue type. The rest looks good.

Should I create a separate query resolver for it because as of now there's no query that asks for venue id? Also, I thought we don't need to fetch organization information of a venue through venue id for the requested features as of now. I hope I didn't miss something. 😅

**Comment by palisadoes** [comment]
That should be OK for now

@EshaanAgg @sumitra19jha Please take another look again

**Comment by wingman47** [comment]
@palisadoes I am getting formatting error in PR Workflow but I don't have any formatting issues in those files. I was modifying the tests for new changes.
image

**Comment by palisadoes** [comment]
Sometimes the workflow detects files you edited due to merge conflicts

**Comment by wingman47** [comment]
@palisadoes @EshaanAgg please review the pull request. All the required changes have been made and I have also added the tests.

File: schema.graphql [hunk 11]
```diff
@@ -211,6 +211,15 @@ input DonationWhereInput {
   name_of_user_starts_with: String
 }
 
+input EditVenueInput {
+  _id: ID!
+  capacity: Int
+  description: String
+  file: String
+  name: String
+  organizationId: ID!
```
Comment (EshaanAgg) [inline schema.graphql:220]: No need for `organizationId` here.

File: schema.graphql [hunk 12]
```diff
@@ -211,6 +211,15 @@ input DonationWhereInput {
   name_of_user_starts_with: String
 }
 
+input EditVenueInput {
+  _id: ID!
```
Comment (EshaanAgg) [inline schema.graphql:215]: Rename to `id`

File: schema.graphql [hunk 13]
```diff
@@ -1207,6 +1221,23 @@ type UsersConnectionResult {
   errors: [ConnectionError!]!
 }
 
+type Venue {
+  _id: ID!
```
Comment (EshaanAgg) [inline schema.graphql:1225]: Rename to `id`

**Review by EshaanAgg — CHANGES_REQUESTED** [review]
You also need to create a field level resolver for `Organization` on the `Venue` model

File: schema.graphql (hunk 11 above)
Comment (wingman47) [reply schema.graphql:220]: `organizationId` is being used here to check here if the user who wants to edit the venue is admin of the organization or not. Should I take any other approach?

Comment (EshaanAgg) [reply schema.graphql:220]: Fetch it from the database. The id is a unique identifier in itself. We can't let the user provide a faulty organizationId or change the organization itself in the venue.

**Comment by palisadoes** [comment]
Please fix the conflicting files

**Comment by palisadoes** [comment]
You have a field named `organizationID` referencing an `organization`.  This negatively impacts readability. Please modify.

image

**Comment by wingman47** [comment]
> You have a field named `organizationID` referencing an `organization`. This negatively impacts readability. Please modify.
> 
> image
> 
> Then you have this, which adds to the inconsistency :
> 
> image

Ok, I will rename it to organization. Will it be okay?
Also, I am unable to write unit test for `updatedVenue is possibly null` in `editVenue` mutation because the `venueId` being using in `findOneAndUpdate` is already verified above so there's no possibillity of `updatedVenue` to be null and I am not allowed to use no null assertion for `updatedVenue` in return due to the workflow rules. Any suggestion to avoid this?

image

**Comment by sumitra19jha** [comment]
This branch is out-of-date with the base branch

**Comment by wingman47** [comment]
> This branch is out-of-date with the base branch

Done
I am unable to increase code cov due to above stated reason

**Comment by palisadoes** [comment]
*NOTE* Read very carefully

1. We just merged this PR which upgraded the prettier package. 
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1840
1. It reformatted over 400 files. This will put your PR at risk of extensive merge conflicts.    
1. I suggest that before you merge your code to the latest upstream, or manually fix these merge conflicts that you:
    1. upgrade your `prettier` in your local branch to the same version
    1. run `prettier`
    
This will help to reduce the number of future merge conflicts for your PR.

**Comment by wingman47** [comment]
@palisadoes I have made all the requested changes.

**Comment by palisadoes** [comment]
@EshaanAgg @sumitra19jha @SiddheshKukade 

PTAL

**Comment by SiddheshKukade** [comment]
@wingman47 please take a pull from origin develop

**Comment by wingman47** [comment]
> @wingman47 please take a pull from origin develop

Done

**Comment by EshaanAgg** [comment]
Please address all the previous conversations left in the previous reviews, and close the ones that you have addressed.

**Comment by wingman47** [comment]
> Please address all the previous conversations left in the previous reviews, and close the ones that you have addressed.

I have addressed all the reviews. 
Also, I am unable to cover a test case in edit Venue because of this reason:


> > You have a field named `organizationID` referencing an `organization`. This negatively impacts readability. Please modify.
> > image
> > Then you have this, which adds to the inconsistency :
> > image
> 
> Ok, I will rename it to organization. Will it be okay? Also, I am unable to write unit test for `updatedVenue is possibly null` in `editVenue` mutation because the `venueId` being using in `findOneAndUpdate` is already verified above so there's no possibillity of `updatedVenue` to be null and I am not allowed to use no null assertion for `updatedVenue` in return due to the workflow rules. Any suggestion to avoid this?
> 
> image

Any suggestion to handle this?

**Comment by wingman47** [comment]
@EshaanAgg Please review the PR

**Review by palisadoes — APPROVED** [review]
